### PR TITLE
Bring back iPXE installation for Flatcar ARM images

### PR DESCRIPTION
**Pull request by dongsupark**
There is on-going work to support Flatcar Linux images for ARM64 architecture. (for example https://github.com/flatcar-linux/coreos-overlay/pull/88)
To make it integrated with Lokomotive, we should bring back the old iPXE installation code, because at the moment there is no way to natively install Flatcar ARM64 images on Packet.

This PR partly reverts https://github.com/kinvolk/lokomotive-kubernetes/pull/28.

**How to test:**

Add these variables to your arm64 workers/controllers:
```
ipxe_script_url = "https://raw.githubusercontent.com/kinvolk/flatcar-ipxe-scripts/arm64-usr/packet.ipxe"
(controller_)type  = "c2.large.arm"
os_arch = "arm64"
os_channel = "alpha"
```
To test it for amd64 only the following variable is needed for workers/controllers:
```
ipxe_script_url = "https://raw.githubusercontent.com/kinvolk/flatcar-ipxe-scripts/amd64-usr/packet.ipxe"
```
You can mix both types in controller and worker pools.

### CODE REVIEW

**Comment by pothos** [comment]
I'm trying out to restore support also for amd64 iPXE booting:

```
diff --git a/packet/flatcar-linux/kubernetes/controllers.tf b/packet/flatcar-linux/kubernetes/controllers.tf
index 0c41ac2..f66913a 100644
--- a/packet/flatcar-linux/kubernetes/controllers.tf
+++ b/packet/flatcar-linux/kubernetes/controllers.tf
@@ -41,7 +41,7 @@ resource "packet_device" "controllers" {
   hostname         = "${var.cluster_name}-controller-${count.index}"
   plan             = "${var.controller_type}"
   facilities       = ["${var.facility}"]
-  operating_system = "${var.os_arch == "arm64" ? "custom_ipxe" : format("flatcar_%s", var.os_channel)}"
+  operating_system = "${var.ipxe_script_url != "" ? "custom_ipxe" : format("flatcar_%s", var.os_channel)}"
   billing_cycle    = "hourly"
   project_id       = "${var.project_id}"
   user_data        = "${element(data.ct_config.controller-ignitions.*.rendered, count.index)}"
diff --git a/packet/flatcar-linux/kubernetes/workers/workers.tf b/packet/flatcar-linux/kubernetes/workers/workers.tf
index d92b0ad..747e8d8 100644
--- a/packet/flatcar-linux/kubernetes/workers/workers.tf
+++ b/packet/flatcar-linux/kubernetes/workers/workers.tf
@@ -3,7 +3,7 @@ resource "packet_device" "nodes" {
   hostname         = "${var.cluster_name}-${var.pool_name}-worker-${count.index}"
   plan             = "${var.type}"
   facilities       = ["${var.facility}"]
-  operating_system = "${var.os_arch == "arm64" ? "custom_ipxe" : format("flatcar_%s", var.os_channel)}"
+  operating_system = "${var.ipxe_script_url != "" ? "custom_ipxe" : format("flatcar_%s", var.os_channel)}"
   billing_cycle    = "hourly"
   project_id       = "${var.project_id}"
   ipxe_script_url  = "${var.ipxe_script_url}"
```


Edit: maybe the gpg key for ARM also should not be used for the installer then…
Edit2: Note: `/opt/installer` also can use the `-s` option for `flatcar-install` instead of a custom logic
Edit3: Result: I ended up in a ipxe booted instance but nothing got installed and the files supposed to be created by ignition where not present, e.g., `/opt/installer`.
Edit4: Ah, the OS is not installed… the wrong ignition template is used!
Edit5: pushed the above patch as commit

**Comment by t-lo** [comment]
Adding @rata and @surajssd for review as this branch is in daily use for the Ampere benchmark project.
This branch allows for provisioning Arm clusters, which is currently not available from Packet. It also allows for provisioning hybrid clusters (i.e. mixed Arm / x86).

**Comment by rata** [comment]
@t-lo let me know when it is ready for review. The first thing that needs to be cleaned up, from a quick look, is the commits and it's reverts. I don't see why we need a patch we later rebase. But I see this is a draft PR, so... let me know when it is ready for review or if I should look at something in particular :)

File: packet/flatcar-linux/kubernetes/workers/cl/install.yaml.tmpl
Comment (rata) [inline packet/flatcar-linux/kubernetes/workers/cl/install.yaml.tmpl]: This ignores the `setup_raid` variable, so those features won't be available on arm without warning

**Comment by pothos** [comment]
It uses the wrong Ignition file and therefore does not install to disk.
Also, the installation should use the `-s` flag for `flatcar-install`.

**Comment by pothos** [comment]
I've switched to the installation ignition template and adapted it to the current needs of the alpha arm64 image.

**Comment by pothos** [comment]
The current Flatcar alpha image does not have the `-s` option for `flatcar-install`, so I went with `-d /dev/sda` as temporary solution. All Packet arm64 servers just have just one SSD anyway.

File: packet/flatcar-linux/kubernetes/workers/cl/install.yaml.tmpl
Comment (pothos) [reply packet/flatcar-linux/kubernetes/workers/cl/install.yaml.tmpl]: The variable is passed to the second ignition and on first boot of the installed OS the RAID will be created as with the native Packet installation.

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (invidian) [inline packet/flatcar-linux/kubernetes/bootkube.tf]: Where those version come from?

File: docs/flatcar-linux/packet.md
Comment (invidian) [inline docs/flatcar-linux/packet.md]: I think there should be one more control variable called `use_ipxe`, so user don't have to explicitly unset that to use Packet images.

Or we could simply use `os_arch` for controlling when to use iPXE.

File: docs/flatcar-linux/packet.md
Comment (invidian) [inline docs/flatcar-linux/packet.md]: We should note, that this only works in iPXE mode.

File: docs/flatcar-linux/packet.md
Comment (invidian) [reply docs/flatcar-linux/packet.md]: Ah, also, it seems, that the documentation is wrong, as actual default value is `""`.

**Review by invidian — CHANGES_REQUESTED** [review]
Mis-clicked the button in previous review. I found some nits.

Also, there seems to be several commits changing same fragment of code multiple times, it would be good to have those changes squashed before we merge. For example fixing "Flatcar Container Linux".

In general LGTM.

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (pothos) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Oh, current version should be `calico/*:v3.10.1`

Comment (invidian) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Actually, I'm talking about all images here. Will we have to always sync it up now when we update bootkube module?

File: docs/flatcar-linux/packet.md
Comment (pothos) [reply docs/flatcar-linux/packet.md]: `os_arch` is independent from iPXE boot

File: docs/flatcar-linux/packet.md
Comment (pothos) [reply docs/flatcar-linux/packet.md]: Thanks, done

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (pothos) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: We can also change the bootkube defaults to the values here.

Comment (pothos) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Except for the `os_arch` suffix, this would need some more changes for bootkube.

Comment (invidian) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Yeah, I think it would be better to propagate that to, I assume `terraform-render-bootkube` module, so we have single source of truth across our code.

Comment (pothos) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Have changed it and will test now.

**Comment by pothos** [comment]
TODO: Use `quay.io/kinvolk/calico-hostendpoint-controller:v0.0.2-ARCH`

**Comment by kosyfrances** [comment]
I created a cluster with an arm worker and noticed that the calico-hostendpoint-controller pod was crashing. 
```
$ kubectl get pods -A                                                
NAMESPACE     NAME                                            READY   STATUS             RESTARTS   AGE
kube-system   calico-hostendpoint-controller-9f95569d-z5czn   0/1     CrashLoopBackOff   9          28m
kube-system   calico-node-nrhxp                               1/1     Running            0          26m
kube-system   calico-node-tfxhf                               1/1     Running            0          28m
....
```
This is not the case when you create a cluster with the previous defaults (i.e without arm).
The logs only give:
```
$ kubectl logs -n kube-system calico-hostendpoint-controller-9f95569d-z5czn
standard_init_linux.go:190: exec user process caused "exec format error"
```
My guess is that the container image for calico-hostendpoint-controller does not have the right architecture for arm machines.

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (pothos) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Works

**Comment by pothos** [comment]
Fixed but with a temporary location in `kailueke/calico-hostendpoint-controller:v0.0.2`
Edit: Moved to `kinvolk/calico-hostendpoint-controller:v0.0.2`

```
NAMESPACE     NAME                                              READY   STATUS    RESTARTS   AGE     IP             NODE                  NOMINATED NODE   READINESS GATES
kube-system   calico-hostendpoint-controller-6f4c97c9db-7j9mq   1/1     Running   0          7m57s   10.2.184.1     k-test-he-worker-0    <none>           <none>
kube-system   calico-node-5brfr                                 1/1     Running   0          7m57s   10.88.72.141   k-test-controller-0   <none>           <none>
```
(k-test-he-worker-0 is c2.large.arm)

File: packet/flatcar-linux/kubernetes/variables.tf
Comment (invidian) [inline packet/flatcar-linux/kubernetes/variables.tf]: Here it should be also mentioned, that this is only for iPXE installation.

**Review by invidian — COMMENTED** [review]
Codewise the PR looks good, however commits are a bit confusing. I didn't check if maybe GitHub is not messing up the order of the commits, I was reviewing it only via GitHub interface.

For the sake of possible debugging/reverting/bisecting in the future, I would re-organize it a bit, so we don't change things back and forth, as this intermediate step is not an intended state.

Here my suggestions for rebasing:

- 895faf3bf4788f9166dd40f0980f0512b33184f7 changes default value of `ipxe_script_url` to new URL, but documentation is changed to `""`. I think this commit should be dropped, as fda83988e886127005116d5d730e52ab62795424 changes it again later. I'd split and fixup to the first commit.

- 6055771ecbfdc5bdbcdd8bc8c118ad87c9d1bf53 adds temp key download. Then ed29e76c3ac7a49daeeb3f1e9c751038a3d2fe87 changes it again.

- 54f03253173d89af410dd203d4bced9835bfcaf5 adds `os_arch` variable and updates bootkube. Update should be perhaps done first in separated commit.

- fda83988e886127005116d5d730e52ab62795424 changes `operating_system` logic. Then it's changed again in d56eaae5f02803e01d4a7fa26ed72540a56598c1.

I would reorganize it to do the following:
- keep "Revert ..." commits as is at the beginning
- update bootkube (should be no-op)
- introduce `os_arch` variable and add everything related to it
- update reverted iPXE scripts, as they rely on `os_arch`

Not sure if I included all required steps, but I hope you get the idea.

On the other hand, if you don't want to spend time on it @pothos and other approve the PR, feel free to ignore me and just merge it.

**Review by kosyfrances — APPROVED** [review]
LGTM

File: packet/flatcar-linux/kubernetes/variables.tf
Comment (pothos) [reply packet/flatcar-linux/kubernetes/variables.tf]: done

**Comment by pothos** [comment]
Squashed some commits.

File: packet/flatcar-linux/kubernetes/variables.tf
Comment (invidian) [inline packet/flatcar-linux/kubernetes/variables.tf:39]: Seems Flatcar Container Linux fix slipped into this commit.

Comment (invidian) [reply packet/flatcar-linux/kubernetes/variables.tf:39]: Also `operating_system` changes doesn't seem to be related to documentation update.

File: packet/flatcar-linux/kubernetes/calico/host-endpoint-controller.yaml
Comment (invidian) [inline packet/flatcar-linux/kubernetes/calico/host-endpoint-controller.yaml:30]: Shouldn't that be 0.0.3 by now?

**Comment by invidian** [comment]
Also merged without waiting for CI :disappointed:

File: packet/flatcar-linux/kubernetes/calico/host-endpoint-controller.yaml
Comment (pothos) [reply packet/flatcar-linux/kubernetes/calico/host-endpoint-controller.yaml:30]: Comes through the other PR that is open.

**Pull request merged**
